eth/stagedsync/stage_tevm.go

```diff
@@ -0,0 +1,344 @@
+package stagedsync
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/c2h5oh/datasize"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/ledgerwatch/erigon/common"
+	"github.com/ledgerwatch/erigon/common/dbutils"
+	"github.com/ledgerwatch/erigon/eth/stagedsync/stages"
+	"github.com/ledgerwatch/erigon/ethdb"
+	"github.com/ledgerwatch/erigon/log"
+	"github.com/ledgerwatch/erigon/metrics"
+	"github.com/ledgerwatch/erigon/params"
+)
+
+var stageTranspileGauge = metrics.NewRegisteredGauge("stage/tevm", nil)
+
+type TranspileCfg struct {
+	db            ethdb.RwKV
+	batchSize     datasize.ByteSize
+	readerBuilder StateReaderBuilder
+	writerBuilder StateWriterBuilder
+	chainConfig   *params.ChainConfig
+}
+
+func StageTranspileCfg(
+	kv ethdb.RwKV,
+	batchSize datasize.ByteSize,
+	readerBuilder StateReaderBuilder,
+	writerBuilder StateWriterBuilder,
+	chainConfig *params.ChainConfig,
+) TranspileCfg {
+	return TranspileCfg{
+		db:            kv,
+		batchSize:     batchSize,
+		readerBuilder: readerBuilder,
+		writerBuilder: writerBuilder,
+		chainConfig:   chainConfig,
+	}
+}
+
+type contract struct {
+	hash []byte
+	code []byte
+}
+
+func transpileBatch(logPrefix string, s *StageState, fromBlock uint64, toBlock uint64, tx ethdb.RwTx, batch ethdb.DbWithPendingMutations, cfg TranspileCfg, useExternalTx bool, quitCh <-chan struct{}) error {
+	done := make(chan struct{})
+	inContracts := make(chan contract, 8192)
+	outContracts := make(chan contract, 8192)
+
+	readWG := new(errgroup.Group)
+	numWorkers := runtime.NumCPU()
+	for i := 0; i < numWorkers; i++ {
+		readWG.Go(func() error {
+			for {
+				select {
+				case evmContract, closed := <-inContracts:
+					if !closed {
+						return nil
+					}
+
+					transpiledCode, err := transpileCode(evmContract.code)
+					if err != nil {
+						return tryError(fmt.Errorf("contract %q cannot be transalated: %w",
+							common.BytesToHash(evmContract.hash).String(), err), done)
+					}
+
+					outContracts <- contract{evmContract.hash, transpiledCode}
+				case <-done:
+					return nil
+				case <-quitCh:
+					return common.ErrStopped
+				}
+			}
+		})
+	}
+
+	logEvery := time.NewTicker(logInterval)
+	defer logEvery.Stop()
+
+	stageProgress := uint64(0)
+	logBlock := stageProgress
+	logTime := time.Now()
+
+	// store translated results
+	writeWG := new(errgroup.Group)
+	writeWG.Go(func() error {
+		var err error
+		for {
+			select {
+			case tevmContract, closed := <-outContracts:
+				if !closed {
+					return nil
+				}
+
+				err = batch.Put(dbutils.ContractTEVMCodeBucket, tevmContract.hash, tevmContract.code)
```

Comment:
Not sure you are allowed to write concurrently into MBDX transaction. Why not make it single threaded first so we can see what performance is? And above.

Reply:
This would probably need to be ETL invocation instead

Reply:
I don't believe it's concurrent. There's only one writing goroutine here.